Update ClientState

Haselnussbomber opened the pull request:
To address issue #2409, this PR adds the following to ClientState:

- A new `ZoneInit` event, fired when the game handles the ZoneInit packet. This essentially restores the previous `TerritoryChanged` behavior, but with additional info.
- A new `MapIdChanged` event, fired whenever `MapId` changes.
- A new `Instance` property, returning the current zone instance used to reduce congestion.
- A new `InstanceChanged` event, fired whenever `Instance` changes.

MidoriKami commented on Dalamud/Game/ClientState/ClientState.cs:
```diff
@@ -428,6 +486,8 @@ internal ClientStatePluginScoped()
     {
         this.clientStateService.TerritoryChanged -= this.TerritoryChangedForward;
+        this.clientStateService.MapChanged -= this.MapChangedForward;
+        this.clientStateService.PublicInstanceChanged -= this.PublicInstanceChangedForward;
```
These need cooresponding `this.clientStateService.MapChanged = null`

statements below to ensure that the objects get GC'd, we had an issue with this in the past which is why we always try to set the delegates to null afterwards.

Haselnussbomber commented:
Added a `ZoneInit` event to bring back the old behavior of `TerritoryChanged` with some more info.
Also I've reworked the properties to raise events when their values change, similar to how `INotifyPropertyChanged` works.
